File: MicroPython/pages/homepage.py
from time import localtime
from gc import mem_free
from graphics.graph_strategy.context_graph_interval import ContextGraphInterval
from graphics import page_elements
from config.config import Display, Colors
from sensors.sensor_manager import SensorManager
from config.env import env_vars


def render_homepage(graph_interval):
    """Renders the homepage with information about the selected sensor and selected graph_strategy interval.

    :param graph_interval: The selected graph_strategy interval. Either 1 (Hourly), 2 (Daily), 3 (Weekly) or 4 (Monthly).
    :return: None
    """

    display = Display()
    graph = ContextGraphInterval()
    graph.strategy = graph_interval
    sensor_manager = SensorManager()

    page_elements.clear_fast()
    display().set_pen(Colors.WHITE)

    page_elements.render_homepage_buttons(graph_interval)
    page_elements.render_nav_arrows(6)
    page_elements.render_sensor_details()
    display().line(101, 16, 101, 120, 3)

    if len(sensor_manager.sensors) == 0:
        display().update()
        return

    active_sensor = sensor_manager.sensors[sensor_manager.active_sensor]
    # Value is fixed here to test on same data every time.
    # graph.render_graph((2024, 2, 15, 3), active_sensor, False)

    if active_sensor:
        current_time = localtime()
        graph.render_graph((current_time[0], current_time[1], current_time[2], current_time[6]), active_sensor, False)

    display().update()

    if env_vars['TEST_HOMEPAGE_MEMORY']:
        return mem_free()

File: C++/pico-boilerplate/pages/homepage.cpp
//
// Created by Petr Jurásek on 04.08.2024.
//

#include "homepage.h"

bool was_rendered = false;

void render_homepage(int graph_interval, datetime_t* last_graph_rendered){
    Display& display = Display::getInstance();
    auto& driver = display.getDriver();
    auto& graphics = display.getGraphics();

    clear_fast();
    graphics.set_pen(Colors::WHITE);

    render_homepage_buttons(graph_interval);
    render_nav_arrows(6);
    render_sensor_details();
    render_thick_separator();

    // If time is set and the difference between the last graph rendered and the current time is greater than 2 minutes
    bool time_set = is_rtc_set();
    datetime_t time_before_graph;
    rtc_get_datetime(&time_before_graph);
    int diff = time_before_graph.min - last_graph_rendered->min;

    if (time_set){
        // Get the current active sensor
        SensorManager& sensor_manager = SensorManager::getInstance();
        Sensor* active_sensor = sensor_manager.getSensor(sensor_manager.getActiveSensor());

        // Set the strategy for the graph interval
        std::unique_ptr<StrategyGraphInterval> strategy;
        if (graph_interval == GraphInterval::DAILY){
            strategy = std::make_unique<ConcreteStrategyDaily>();
        } else if (graph_interval == GraphInterval::WEEKLY){
            strategy = std::make_unique<ConcreteStrategyWeekly>();
        } else if (graph_interval == GraphInterval::MONTHLY){
            // strategy = std::make_unique<ConcreteStrategyMonthly>();
        }

        if (graph_interval != GraphInterval::MONTHLY){
            ContextGraphInterval setStrategy(std::move(strategy));
            setStrategy.renderGraph(get_current_datetime(), active_sensor, false);
        }

        was_rendered = true;
        rtc_get_datetime(last_graph_rendered);
    } else {
        graphics.set_pen(Colors::RED);
        graphics.text("NO GRAPH", Point{120, DISPLAY_HEIGHT/2-10}, 250, 2);
    }

    driver.update(&graphics);
}